src/tokenizer.rs

```diff
@@ -912,18 +926,21 @@ impl<'a> Tokenizer<'a> {
     fn tokenize_identifier_or_keyword(
         &self,
         ch: impl IntoIterator<Item = char>,
-        chars: &mut State,
+        chars: &mut State<'a>,
     ) -> Result<Option<Token>, TokenizerError> {
         chars.next(); // consume the first char
-        let ch: String = ch.into_iter().collect();
-        let word = self.tokenize_word(ch, chars);
+                      // Calculate total byte length without allocating a String
+        let consumed_byte_len: usize = ch.into_iter().map(|c| c.len_utf8()).sum();
```

Comment:
I wonder if we can instead replace the `ch` parameter to this `tokenize_identifier_or_keyword` function with the actual `consumed_byte_len: usize`?

If I'm understanding the intent/requirement of this change correctly the caller of this function should have that value on hand given that we seem to require that `ch` contains the preceeding characters in the stream?

The current flow was initially unclear to me that the caller passes in an iterator of items, whose contents we did not use here, and it required digging further into `tokenize_word` to realised why that was the case.

Reply:
agree , move it out . 
The downside is that callers now need to calculate the UTF-8 byte length of their characters.